### PR TITLE
[Reviewer: Andy] recv_file_descriptor return -1 on error

### PR DESCRIPTION
**Pull request by johadalin**
This is to fix issue #408 

This should stop it returning a 0 on 'error', which would happen, according to http://linux.die.net/man/2/recvmsg, if the peer has performed an orderly shutdown. This fixes an error where listing a sas in shared_config, but not restarting clearwater-infrastructure to add it to the whitelist, causes sprout to become unresponsive and fall into a restart loop, caused by receiving file descriptor '0' for a socket.

I am not sure what essentially copying https://github.com/Metaswitch/sas-client/blob/clearwater/source/sas.cpp#L403 across would do, as i'm not quite sure why the differences have evolved, or what ramifications changes would make, so this PR is just to fix the bug that was causing sprout to become unresponsive and be shut down by monit.

The code still builds, and a previously faulty deployment has been upgraded and now doesn't hit the restart loop, so it appears to be fixed. Let me know if you think we still need to update this function to be more like the old code.

### CODE REVIEW

**Comment by bossmc** [comment]
Hmmmm... Beyond the better commenting of the `namespace_hop` version, I prefer almost everything about the `sas-client` implementation, so it would be good to move the code over (and merge in the comments).

Your specific fix looks fine.

**Comment by johadalin** [comment]
Do you want any more than this pulled across?

File: src/namespace_hop.cpp [hunk 1]
```diff
@@ -52,11 +52,12 @@ static int recv_file_descriptor(int socket)
 {
   // Buffer to hold the dummy "actual data" we receive
   char data[1] = {0};
-  struct iovec iov = {.iov_base = data,
-                      .iov_len = sizeof(data)};
+
+  struct iovec iov[1];
```
Comment (bossmc) [inline src/namespace_hop.cpp]: This needs the `= {0}` (since you're not `memset`-ing to 0).

File: src/namespace_hop.cpp [hunk 2]
```diff
@@ -65,17 +66,19 @@ static int recv_file_descriptor(int socket)
...
   if (res <= 0)
   {
     TRC_WARNING("Failed to retrieve cross-namespace socket - recvmsg returned %d (%d %s)\n", res, errno, strerror(errno));
-    return res;
+    return -1;
   }
 
   // Iterate through control message to find if there is a SCM_RIGHTS entry containing file descriptors
+  struct cmsghdr *control_message = NULL;
+
```
Comment (bossmc) [inline src/namespace_hop.cpp]: Extra blank line.

**Comment by bossmc** [comment]
Two comments, otherwise looks good.

**Pull request merged**